packages/devtools/lib/src/debugger/debugger_state.dart

```diff
@@ -292,3 +290,18 @@ class DebuggerState {
     _breakpoints.add(isolate.breakpoints);
   }
 }
+
+class _Property<T> {
```

Comment:
Lets use ValueNotifier from Flutter instead. You can copy and paste the code from flutter to the fake_flutter library.
That will also set us up better for the migration to flutter_web.

@DaveShuckerow correct me if this isn't the right class from flutter to use in this case.
https://api.flutter.dev/flutter/foundation/ValueNotifier-class.html

Reply:
`ValueNotifier` does not expose a way to get a `Stream` out.

I can either make a change to this class API and expose as a `ValueListenable<T>` instead of a `Stream<bool>`. Or I could stick in some boilerplate to expose as a `Stream`...

Reply:
ValueNotifier is the Flutter way of doing this, but note that Flutter architecture avoids streams, and ValueNotifier doesn't expose one.

This looks like it would require changing the API of DebuggerState to support.

I personally am a fan of avoiding streams for state management and just having a notifier, but that might be a larger change.

Reply:
It's a smaller change to this code to continue to be stream based, so I would prefer that.

Reply:
Started on the deeper change before seeing this. We can use it as a platform for discussion.

Switching the interface to expose as `ValueListenable` wasn't too huge. I have concerns about the semantics around acting on the the initial value when setting up the listeners. In some places I made the change to act on the initial value, in other places that seemed a little involved and I didn't do it.


I also had to change some of the semantics around `breakpoints` to make copies of the list rather than modify in place. `ValueNotifier` doesn't have (supported) semantics for notifying listeners despite the value not changing equality.

Reply:
@devoncarew - Does the current change including copying over `ValueNotifier` look OK to you?

Reply:
It looks like every place where we use `.addListener()`, we duplicate the code for what to do when the value changes (once to assign the current value, and once when the value changes). Those two copies can now get out of date over time as people work on the codebase.

Avoiding code duplication like that is the reason why lodash and other frameworks fire an event when a listener is added - you only have to write the code that reacts to an event.

We could add a `fireOnListenerAdd` param to the constructor of ValueNotifier. It would let us keep this change - to remove the dep on `rxdart` - without introducing code duplication.

Reply:
I don't have bandwidth to drive a change like that in flutter.

I'd be fine making it locally, which would make it a blocker to migrate to flutter later.


I'm worried that we're optimizing for behavior we don't have fully fleshed out. No tests are written that catch this. The existing code doesn't give me a lot of confidence - the `concat` bug has me worried. We handle things like `_updatePauseButton()` both it its own click handler, as well as when the debugger reports a change in the pause state - do we have a race condition there?

In my opinion we'd be safer if we had a more clear separation between code handling the initial state setup and code handling updates - relying on update handling code for initial state scares me.

Reply:
I would also prefer fireOnListenerAdd but I care more about us not forking ValueNotifier as I want to be able to easily swap out the real package:flutter version.

Reply:
>  Those two copies can now get out of date over time as people work on the codebase.

For everything other than assignment to another field I broke things out into a separate function so there is only 1 copy to update.